src/main/scala/org/ergoplatform/modifiers/history/PoPowAlgos.scala

```diff
@@ -0,0 +1,80 @@
+package org.ergoplatform.modifiers.history
+
+import org.ergoplatform.mining.difficulty.RequiredDifficulty
+import org.ergoplatform.settings.Constants
+import scorex.util.{ModifierId, bytesToId, idToBytes}
+
+/**
+  * A set of utilities for working with PoPoW security protocol.
+  */
+object PoPowAlgos {
+
+  val InterlinksFieldsPrefix: Byte = 0x01
+
+  /**
+    * Computes interlinks vector for the next level after `prevHeader`.
+    */
+  @inline def updateInterlinks(prevHeader: Header, prevInterlinks: Seq[ModifierId]): Seq[ModifierId] = {
+    if (!prevHeader.isGenesis) {
+      assert(prevInterlinks.nonEmpty, "Interlinks vector could not be empty in case of non-genesis header")
+      val genesis = prevInterlinks.head
+      val tail = prevInterlinks.tail
+      val prevLevel = maxLevelOf(prevHeader)
+      if (prevLevel > 0) {
+        (genesis +: tail.dropRight(prevLevel)) ++ Seq.fill(prevLevel)(prevHeader.id)
+      } else {
+        prevInterlinks
+      }
+    } else {
+      Seq(prevHeader.id)
+    }
+  }
+
+  /**
+    * Packs interlinks into extension key-value format.
+    */
+  @inline def packInterlinks(links: Seq[ModifierId]): Seq[(Array[Byte], Array[Byte])] = {
+    def loop(rem: Seq[(ModifierId, Int)],
+             acc: Seq[(Array[Byte], Array[Byte])]): Seq[(Array[Byte], Array[Byte])] = {
+      rem match {
+        case (headLink, idx) :: _ =>
+          val duplicatesQty = links.count(_ == headLink)
+          val filed = Array(InterlinksFieldsPrefix, idx.toByte) -> (duplicatesQty.toByte +: idToBytes(headLink))
+          loop(rem.drop(duplicatesQty), acc :+ filed)
+        case Nil =>
+          acc
+      }
+    }
+    loop(links.zipWithIndex, Seq.empty)
+  }
+
+  /**
+    * Unpacks interlinks from key-value format of extension.
+    */
+  @inline def unpackInterlinks(fields: Seq[(Array[Byte], Array[Byte])]): Seq[ModifierId] = {
+    fields
+      .filter(_._1.headOption.contains(InterlinksFieldsPrefix))
+      .foldLeft(Seq.empty[ModifierId]) { case (acc, (_, v)) =>
+        assert(v.lengthCompare(Constants.ModifierIdSize + 1) == 0)
```

Comment:
Using assert (again, better require to avoid confusions) with no catching to catch events which should never happen (if the implementation is done properly) is okay, however, when external data is being parsed:
1. All the checks should be non-elidable
2. Possible faults should be clearly visible to the function callers, i.e. the function should return Try[...]

Also, please write a test for parsing wrong interlinks bytes.